File: 01-intro/03-basic-combinatorial-algorithms/03-combinations/iterative/combination-iterative-with-repetition.py
from pprint import pprint as print

class Combination(object):
    elements = []
    elements_count = 0
    slots = 0
    index_buffer = []
    combinations = []

    def __init__(self, elements: list, slots: int):
        self.elements = elements
        self.elements_count = len(elements)
        self.slots = slots

    def compute_slot(self, slot_index: int, combination_index: int):
        index = (combination_index // (self.elements_count ** slot_index)) % \
               self.elements_count

        prev_index = self.index_buffer[-1] if len(self.index_buffer) > 0 else -1

        if prev_index != [] and prev_index > index:
            raise Exception()

        self.index_buffer.insert(0, index)
        return self.elements[index]

    def compute_combination(self, combination_index: int) -> list:
        slot_range = range(self.slots - 1, -1 ,-1)
        combination = []

        for slot_index in slot_range:
            combination.append(self.compute_slot(slot_index, combination_index))

        self.index_buffer = []

        return combination

    def compute(self):
        index_range = range(0, self.elements_count ** self.slots)

        for combination_index in index_range:
            try:
                self.combinations.append(
                    self.compute_combination(combination_index)
                )
            except:
                self.index_buffer = []

        return self


print(Combination(["a", "b", "c", "d"], 2).compute().combinations)
# Combination(["a", "b", "c", "d", "e"], 3).compute()

File: 01-intro/03-basic-combinatorial-algorithms/03-combinations/iterative/combination-iterative-with-repetition.ex
defmodule Combination do
  def pow(base, p), do: :math.pow(base, p) |> round()

  def compute_combination(slot_index, _, _, {_, _, _}) when slot_index == -1 do
    []
  end

  def compute_combination(
        slot_index,
        combination_index,
        index_buffer,
        {elements, elements_count, slots}
      ) do
    index = combination_index |> div(pow(elements_count, slot_index)) |> rem(elements_count)
    prev_index = Enum.at(index_buffer, -1, -1)

    if prev_index != [] and prev_index >= index do
      raise "not valid"
    else
      head = [Enum.at(elements, index)]

      tail =
        compute_combination(
          slot_index - 1,
          combination_index,
          index_buffer ++ [index],
          {elements, elements_count, slots}
        )

      head ++ tail
    end
  end

  def compute(elements, slots) do
    elements_count = Enum.count(elements)
    indexes = 0..(pow(elements_count, slots) - 1)

    for combination_index <- indexes do
      try do
        compute_combination(slots - 1, combination_index, [], {elements, elements_count, slots})
      rescue
        RuntimeError -> []
      end
    end
    |> Enum.filter(fn combination -> not Enum.empty?(combination) end)
  end
end

# Combination.compute(["a", "b", "c", "d"], 2)
# Combination.compute(["a", "b", "c", "d", "e"], 3)
IO.inspect(Combination.compute(["a", "b", "c", "d"], 2))
# IO.inspect(Combination.compute(["a", "b", "c", "d"], 2))
# IO.inspect(Combination.compute([0, 1, 2, 3, 4], 2))
# IO.inspect(Combination.compute([0, 1, 2], 3))
# IO.inspect(Combination.compute([0, 1, 2, 3, 4], 3))
# IO.inspect(Combination.compute(["a", "b", "c"], 3))
# Combination.compute(["a", "b", "c"], 2)
